algorithms/nautilus/grammartec/mutator.cpp

```diff
@@ -0,0 +1,373 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2022 Ricerca Security
+ * 
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+/**
+ * @file mutator.cpp
+ * @brief Tree mutation engine
+ * @author Ricerca Security <fuzzuf-dev@ricsec.co.jp>
+ *
+ * @details This file implements the mutation and minimization
+ *          algorithm of Nautilus.
+ *          It contains the subtree minimization, recursive minimization,
+ *          random mutation, rules mutation, random recursive mutation,
+ *          and splicing mutation.
+ */
+#include "fuzzuf/algorithms/nautilus/grammartec/mutator.hpp"
+#include "fuzzuf/utils/common.hpp"
+#include "fuzzuf/utils/random.hpp"
+
+
+namespace fuzzuf::algorithm::nautilus::grammartec {
+
+/**
+ * @fn
+ * Minimize a tree with constraints by tester satisfied.
+ * @brief Subtree Minimization
+ * @param (tree) Tree
+ * @param (bits) Set of bits to be passed to tester
+ * @param (ctx) Context
+ * @param (start_index) Beginning of node ID
+ * @param (end_index) End of node ID
+ * @param (tester) Tester function to check if a tree satisfies constraints
+ * @return Returns true if minimization is complete, otherwise false
+ *
+ * @details Subtree minimization makes subtrees as short as possible
+ *          while still triggering the same paths of the original
+ *          testcase.
+ */
+bool Mutator::MinimizeTree(Tree& tree,
+                           std::unordered_set<size_t>& bits,
+                           Context& ctx,
+                           size_t start_index, size_t end_index,
+                           FTester& tester) {
+  size_t i = start_index;
+
+  /* For each nonterminal, we generate the smallest possible subtree */
+  while (i < tree.Size()) {
+    NodeID n(i);
+    NTermID nt = tree.GetRule(n, ctx).Nonterm();
+
+    /* Make sure it's the smallest */
+    if (tree.SubTreeSize(n) > ctx.GetMinLenForNT(nt)) {
+      /* We sequentially replace each node's subtree with the smallest one */
+      _scratchpad.GenerateFromNT(nt, ctx.GetMinLenForNT(nt), ctx);
+      if (std::optional<Tree> t = Mutator::TestAndConvert(
+            tree, n, _scratchpad, NodeID(0), ctx, bits, tester
+          )) {
+        /* If the new transitions are still triggered, replace the tree */
+        tree = t.value();
```

Comment:
Here, I would prefer `std::move(t.value())`.
You need to add move constructor. But it can be done soon since you have almost the same constructor.

Reply:
Fixed in [2bf662f](https://github.com/fuzzuf/fuzzuf/pull/23/commits/2bf662f641b85151bbb79f7a6cf99a62230d868c)

Reply:
I think this is not enough.
It's possible that t.value() will be copied regardless of std::move because the default move constructor and operator is defined only in limited situations.
So please define move operator explicitly.

Reply:
Fixed in [77dc662](https://github.com/fuzzuf/fuzzuf/pull/23/commits/77dc662bf898f43abc8ecc92ec76ae6a602186df)